deeplake/core/tensor.py

```diff
@@ -1365,3 +1375,88 @@ def creds_key(self):
     def invalidate_libdeeplake_dataset(self):
         """Invalidates the libdeeplake dataset object."""
         self.dataset.libdeeplake_dataset = None
+
+    def create_vdb_index(
+        self, id: str, distance: Union[DistanceType, str] = DistanceType.L2_NORM
+    ) -> Indexer:
+        self.storage.check_readonly()
+        if self.meta.htype != "embedding":
+            raise Exception(f"Only supported for embedding tensors.")
+        if not self.dataset.libdeeplake_dataset is None:
+            ds = self.dataset.libdeeplake_dataset
+        else:
+            from deeplake.enterprise.convert_to_libdeeplake import (
+                dataset_to_libdeeplake,
+            )
+
+            ds = dataset_to_libdeeplake(self.dataset)
+        ts = getattr(ds, self.meta.name)
+        from indra import api  # type: ignore
+
+        if type(distance) == DistanceType:
+            distance = distance.value
+        self.meta.add_vdb_index(id=id, type="hnsw", distance=distance)
+        try:
+            index = api.vdb.generate_index(ts, index_type="hnsw", distance_type=distance)
+            b = index.serialize()
+            commit_id = self.version_state["commit_id"]
+            self.storage[get_tensor_vdb_index_key(self.key, commit_id, id)] = b
+            self.invalidate_libdeeplake_dataset()
+        except:
+            self.meta.remove_vdb_index(id=id)
+            raise
+        return index
+
+    def delete_vdb_index(self, id: str):
+        self.storage.check_readonly()
+        if self.meta.htype != "embedding":
+            raise Exception(f"Only supported for embedding tensors.")
+        commit_id = self.version_state["commit_id"]
+        self.storage.pop(get_tensor_vdb_index_key(self.key, commit_id, id))
+        self.meta.remove_vdb_index(id=id)
+        self.invalidate_libdeeplake_dataset()
+        self.storage.flush()
+
+    def regenerate_vdb_indexes(self):
+        is_embedding = self.htype == "embedding"
+        has_vdb_indexes = hasattr(self.meta, "vdb_indexes")
+        vdb_index_ids_present = len(self.meta.get_vdb_index_ids()) > 0
+
+        if is_embedding and has_vdb_indexes and vdb_index_ids_present:
+            index_ids = self.meta.get_vdb_index_ids()
+            for id in index_ids:
+                # Delete the index.
+                self.delete_vdb_index(id)
+                # Recreate it back.
+                self.create_vdb_index(id)
+
+    def load_vdb_index(self, id: str) -> Indexer:
```

Comment:
I think this is not correct typing, this should an indra object. But we can't return type in indra format because it will raise an exception, so I suggest to skip it